src/rewrite_quantization.cpp

```diff
@@ -0,0 +1,95 @@
+#include <migraphx/rewrite_quantization.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/tune_axis.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/shape.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+void apply_quantizelinear(module& m, instruction_ref ins)
+{
+    auto x       = ins->inputs()[0];
+    auto y_scale = ins->inputs()[1];
+
+    auto quant_type   = ins->get_shape().type();
+    int64_t max_quant = 0;
+    int64_t min_quant = 0;
+    ins->get_shape().visit_type([&](auto as) {
+        max_quant = as.max();
+        min_quant = as.min();
+    });
+    auto max_arg = m.add_literal(static_cast<int>(max_quant));
+    auto min_arg = m.add_literal(static_cast<int>(min_quant));
+
+    auto div            = m.insert_instruction(ins, make_op("div"), x, y_scale);
```

Comment:
There is no guarantee that x and y_scale are of the same type, we need a check and type conversion here. We also need to add a unit test for that.

Reply:
The implementation can be something like:

```
    auto inputs  = ins->inputs();
    auto scale = inputs[1];
    auto scale_type = scale->get_shape().type();
    if(inputs[0]->get_shape().type() != scale_type)
    {
        inputs[0] =
            m.insert_instruction(ins, make_op("convert", {{"target_type", scale_type}}), inputs[0]);
    }
    auto div            = m.insert_instruction(ins, make_op("div"), inputs[0], scale);
    auto add_zero_point      = m.insert_instruction(ins, make_op("round"), div);

    if(inputs.size() == 3)
    {
        auto zero_point = inputs[2];
        if(scale_type != zero_point->get_shape().type())
        {
            zero_point = m.insert_instruction(
                ins, make_op("convert", {{"target_type", scale_type}}), zero_point);
        }
        add_zero_point = m.insert_instruction(ins, make_op("add"), add_zero_point, zero_point);
    }

    auto s           = add_zero_point->get_shape();
    const auto& lens = s.lens();
    std::vector<int64_t> out_lens(lens.begin(), lens.end());
    std::vector<int> min_data(s.elements(), min_val);
    std::vector<int> max_data(s.elements(), max_val);
    auto min_arg = m.add_literal(literal(s, min_data));
    auto max_arg = m.add_literal(literal(s, max_data));

    auto saturated = m.insert_instruction(ins, make_op("clip"), add_zero_point, min_arg, max_arg);
    m.replace_instruction(ins, make_op("convert", {{"target_type", type}}), saturated);
```